### PR TITLE
One-Way Windows

### PR DESCRIPTION
**Pull request by Shadowmech88**
A window can now be turned into a one-way window by using a sheet of plastic on it.
The plastic can be pried off with a crowbar, turning them back into regular windows.
One-way windows obstruct a player's view from one side, but allow them to see through from the other.

One-way windows block night vision goggles and thermals, but not X-ray vision.
<strike>They also don't block sound and, while I would like them to, I'm not sure how to go about that.</strike>
They also don't prevent right-clicking to see the contents of a tile, and I am again not sure how to fix that.

one_way_window

The window's dir affects the directions from which it blocks vision, even for full windows. When building one-way windows, face the window toward the direction from which you want it to be transparent.
Full windows have a new verb that allows players to manually set their direction, so that they can access the window's diagonal directions, since this PR gives diagonal windows an actual gameplay difference from cardinal windows. Window smoothing doesn't seem to be affected.

For mapping purposes, regular windows can be made into one-way windows by toggling their one_way var.

I'll add these to the Interrogation room once they're merged.

### CODE REVIEW

**Comment by ghost** [comment]
I wish we had proper tinted windows, but this will do just fine for now

**Comment by Shadowmech88** [comment]
You mean windows that adjust visibility depending on the light level on either side?

I imagine that would be a much greater challenge.

**Comment by ghost** [comment]
Nah, I mean recycling the old tinted window type that I left aside when refactoring windows and that nowadays only exists in Interrogation (proper) and the Bridge

Doesn't need to do anything extremely fancy, although one-way vision when in pane form would certainly make it a lot more interesting

**Comment by Shadowmech88** [comment]
I purposefully avoided making it a subtype so that you could make _any_ window a one way window, like plasma windows for example.

**Comment by ghost** [comment]
Right, would be nice to make sure this system is fully integrated though, and that the old tinted windows use this from now on

**Comment by Shadowmech88** [comment]
I could make that change when I add these to the maps, if you'd like me to.
Currently I've left tinted windows alone in case someone wants a completely opaque window, for whatever reason.

**Comment by clusterfack** [comment]
Seems like it would be costly as fuck tbh

**Comment by invoid9** [comment]
Are these the kind of one way windows that work by the opaque side being mirrors? If so, I wanna be able to do my hair in a one way window and have people on the other side laugh at me to each other. 

10/10 roleplay experience right there

**Comment by ihadtoregisterforthis** [comment]
I would have expected the other side of the window to be opaque.
And how does standing on the same tile as the window work?

**Comment by ComicIronic** [comment]
Here's a fun factoid: this won't work if you put a mob in a locker. Has all the same problems as the Parallax update.

**Comment by clusterfack** [comment]
Yet ANOTHER reasaon we need http://www.byond.com/forum/?post=2085576

Someone bug lummox about it again

File: code/game/objects/structures/fullwindow.dm [hunk 1]
```diff
@@ -58,6 +58,39 @@
 		icon_state = "[base_state][junction]"
 		return
 
+/obj/structure/window/full/verb/set_direction() //Full windows get this because it's possible for them to face diagonally
```
Comment (ComicIronic) [inline code/game/objects/structures/fullwindow.dm:61]: I really do not support using verbs. They require a huge hodgepodge of boilerplate to get working correctly - for example, this is going to be ghost-exploitable.

I'd much rather you put spinning an unanchored window into the attack_hand for it and do it like that instead.

Comment (Shadowmech88) [inline code/game/objects/structures/fullwindow.dm:61]: > for example, this is going to be ghost-exploitable

It'd be easy to just add a check to see if the user is living and able.

**Comment by Shadowmech88** [comment]
@ihadtoregisterforthis

If you stand on the same tile as the window, it doesn't work.

File: code/game/objects/structures/fullwindow.dm (hunk 1 above)
Comment (ComicIronic) [inline code/game/objects/structures/fullwindow.dm:61]: Which is boilerplate. Would you also remember the restrained check?

Comment (Shadowmech88) [inline code/game/objects/structures/fullwindow.dm:61]: I went ahead and made a proc that checks whether a mob is living, conscious, and unrestrained, and used that.

Comment (9600bauds) [inline code/game/objects/structures/fullwindow.dm:61]: I already did that
It's called "incapacitated()"

**Comment by clusterfack** [comment]
But yeah calling view() everytime a client moves is really expensive. I mean, if you merged this with the system parallax uses and both called view() at least the expense wouldn't be going to waste? 

It's hard to say since our cpu usage bottomed out after 510, it's not like we are hurting for it at the moment, so a feature like this is alright I hesitantly suppose.

If you do agree with me use the update client hook code I put in, change the hud used part to do a more general client movement update that calls both the parallax update and this depending on what is in view.

**Comment by Shadowmech88** [comment]
I am completely willing to do that.

But just for fun, pretend that I know very little about `client` and then tell me how to do that.

**Comment by clusterfack** [comment]
I added a new client hook in my parallax pr. It finds clients that needs to be updated on move() and updates their parallax. Make that point to a client proc instead that checks view() and updates both parallax and this depending on the visibility circumstances.

**Comment by clusterfack** [comment]
https://github.com/d3athrow/vgstation13/blob/Bleeding-Edge/code/game/atoms_movable.dm#L327

**Comment by Shadowmech88** [comment]
Alright, I'll try.

On an unrelated note, do you know how I could turn the "darkness" created by these windows into _actual_ darkness? Currently it's just manually drawing black squares over the tile, which doesn't prevent you from hearing people through the glass or right-clicking tiles on the other side of it.

**Comment by clusterfack** [comment]
No there's no way to make 'real darkness' that isn't hardcoded into byond. 

However you could theoretically prevent hearing and visible messages by passing the source as an argument through visible message and hear and then checking its location against your mobs obscured turfs list.

**Comment by Probe1** [comment]
That's funny I was just thinking about this last week and how it'd be a nice use for electricity.  Good job.

**Comment by Shadowmech88** [comment]
I've made an attempt to merge the window update code with the parallax update code, but I don't know if I did it properly.

Nothing seems to be broken, at least.

File: code/modules/client/client procs.dm [hunk 2]
```diff
@@ -432,3 +432,53 @@
 	if(!(colour_to.len))
 		colour_to = default_colour_matrix
 	animate(src, color=colour_to, time=time, easing=SINE_EASING)
+
+/client/proc/update_special_views()
+	var/list/vi = view(world.view,get_holder_at_turf_level(mob))
```
Comment (clusterfack) [inline code/modules/client/client procs.dm]: Now you're doing view() and range(), just have parallax done by view()

**Comment by FudgePucker** [comment]
Why does the one way window block more thermal energy than 5 rwalls?

Was blocking thermals intended or a side effect?

**Comment by Shadowmech88** [comment]
@FudgePucker 
It's a side-effect and I don't know how to fix it, since the window's darkness isn't actual darkness.

File: code/modules/client/client procs.dm (hunk 2 above)
Comment (Shadowmech88) [inline code/modules/client/client procs.dm:437]: How's it look now?

**Comment by Shadowmech88** [comment]
Alright, so now the obscurities created by one-way windows block speech and visible_messages from being seen by mobs with clients.
I don't _think_ the way I did it will have any big impact on performance, but I don't know enough to be sure.

Unfortunately clientless hearers like intercoms and tape recorders and whatnot can hear through one-way windows since not having a client means they don't have an obscured turfs list. I'm pretty sure that calling view() on EVERY hearer, including clientless ones, every time anything is said or emoted, just to determine whether the atom hearing it has a one-way window in its view, would be too expensive for the small benefit it would provide.

**Comment by Shadowmech88** [comment]
So is this good to go?

**Pull request merged**